### PR TITLE
[DistMuon] Split BlockShard blocks into matrix segments

### PR DESCRIPTION
**Pull request by JavaZeroo**
## Summary

Part 1 of #4692. MLA fuses several projections into one parameter with different row counts per piece: `wq_b` is per head `[q_nope; q_rope]`, `wkv_b` per head `[k_nope; v]`, `wkv_a` one block `[kv_latent; k_rope]`. Kimi runs Newton-Schulz per logical projection, but `BlockShard` can only cut equal-sized blocks, so there is no way to express `split([nope, v])` today.

This lets `BlockShard.block_size` be a tuple: a block made of contiguous matrices with different row counts. The block is still `sum(block_size)` rows and is never split between ranks; the pieces only tell DistMuon where one matrix ends and the next begins.

```python
ComputeLayout(
    shardings_by_mesh_axis={"dp_shard": BlockShard(dim=0, block_size=(nope, v))},
)
```

- Transport is unchanged: `BlockShard` still moves whole blocks of `nope + v` rows, so the redistribution plan does not know about the pieces. Making each piece its own unit would let a rank's local rows start mid-head and break the `[M, R, C]` batched view.
- At compute time each segment is a narrowed `[M, rows_i, C]` view of the block batch; NS runs once per segment and writes back through the view.
- Segments have different shapes, so the aspect-ratio lr adjustment (`match_rms_adamw` etc.) is applied per segment inside `_compute_update`. It cannot be done in `_apply_update` because the storage shard that applies the update may not hold complete blocks.
- An integer `block_size` behaves exactly as before.

Only `BlockShard` is touched, so this is independent of #4531 and of the w13 layout change in #4676.

The second commit switches the Kimi K2.7 recipe to it: `wq` / `wq_b` per head `(128, 64)`, `wkv_b` per head `(128, 128)`, `wkv_a` one block `(512, 64)`. Today K2.7 runs NS on the whole `[k_nope; v]` block per head and on the whole `wkv_a` matrix, which couples projections that Kimi's implementation keeps separate (https://kexue.fm/archives/11126/comment-page-2#comment-28601). This changes K2.7 numerics on purpose; there are no loss goldens for the K2.5 Muon recipes, only the functionality integration tests. The K3 recipe (#4596) uses the same layouts.

## Test plan

- `tests/unit_tests/gpu/flex_shard/test_dist_muon.py::test_segments_match_independent_matrix_updates`: 2-GPU, one block per rank with `block_size=(3, 2)`, compared against `torch.optim.Muon` on the four split matrices with `match_rms_adamw`. I only have one GPU locally, so this one relies on CI.
- Single-GPU check with the K2.5 debugmodel shapes (`wq [3072, 256]`, `wkv_a [576, 256]`, `wkv_b [4096, 512]`): DistMuon step matches per-matrix `torch.optim.Muon` to 5e-10; against whole-block NS the difference is 3e-4 to 4e-4 on updates of 7e-4.
- Single-GPU check with the K3 debugmodel shapes (`wq_b [1536, 512]`, `wkv_a [288, 1024]`, `wkv_b [2048, 256]`) and an unequal-head-dim variant: DistMuon step matches per-matrix `torch.optim.Muon` to 4.7e-10 for `wkv_a` / `wkv_b`; `wq_b` differs by 7e-6, which is bf16 bmm-vs-mm accumulation order in NS (with a single block it is 1.2e-10), not layout.

### CODE REVIEW

**Review by weifengpy — COMMENTED** [review]
> but BlockShard can only cut equal-sized blocks, so there is no way to express split([nope, v]) today

I am checking motivation from PR description - what do you think of improving blockshard to express non-uniform length? segments in ComputeLayout feels like a downsteam bandit because of blockshard contraints

**Comment by JavaZeroo** [comment]
> > but BlockShard can only cut equal-sized blocks, so there is no way to express split([nope, v]) today
> 
> I am checking motivation from PR description - what do you think of improving blockshard to express non-uniform length? segments in ComputeLayout feels like a downsteam bandit because of blockshard contraints

sure, `BlockShard.block_size` now accepts a tuple and ComputeLayout.num_rows_per_segment is removed

**Review by weifengpy — COMMENTED** [review]
this PR is very valuable! i didn't know we have incorrect numerics. will spend more time to figure out what's needed to land unequal-sized BlockShard

Image

**Comment by weifengpy** [comment]
related: https://github.com/NVIDIA-NeMo/Emerging-Optimizers/issues/31#issuecomment-3372813575

File: torchtitan/models/kimi_k2_7/config_registry.py [hunk 1]
```diff
@@ -329,7 +342,7 @@ def _dist_muon_optimizer(
     )
     attention_shardings = {
         **query_shardings,
-        "wkv_a": owned,
+        "wkv_a": kv_latent_and_rope,
```
Comment (shuhuayu) [inline torchtitan/models/kimi_k2_7/config_registry.py:345]: this helps splitting logical matrices, but all blockshard will place `wkv_a` on the first rank, since all heads share this tensor per layer. will have a load balancing problem.

**Comment by JavaZeroo** [comment]
> this PR is very valuable! i didn't know we have incorrect numerics. will spend more time to figure out what's needed to land unequal-sized BlockShard
> 
> Image

Thanks for the confirmation. I spent a lot of time last year on Muon, mostly on numerical correctness and distributed performance in another training stack, so I'd like to keep contributing in torchtitan.

**Comment by weifengpy** [comment]
@JavaZeroo, I added you as co-author in https://github.com/pytorch/torchtitan/pull/4799
* BlockShard with variable length is core piece in FlexShard, not only DistMuon: https://github.com/meta-pytorch/flex_shard
* I summarized delta below. most importantly, we need to be able to shard fused param correctly at the boundary of a block
* your PR still have extra things like scales each segment’s direction by segment_ratio / block_ratio. are you intersted in rebase?

Screenshot 2026-09-20 at 11 38 31 PM

**Comment by JavaZeroo** [comment]
> @JavaZeroo, I added you as co-author in #4799
> 
> * BlockShard with variable length is core piece in FlexShard, not only DistMuon: https://github.com/meta-pytorch/flex_shard
> * I summarized delta below. most importantly, we need to be able to shard fused param correctly at the boundary of a block
> * your PR still have extra things like scales each segment’s direction by segment_ratio / block_ratio. are you intersted in rebase?
> 
> Screenshot 2026-09-20 at 11 38 31 PM

Thanks for the co-author! I'll rebase onto your PR and continue with the remaining Muon work on top of it.

**Comment by weifengpy** [comment]
> > * your PR still have extra things like scales each segment’s direction by segment_ratio / block_ratio. are you intersted in rebase?

I absorbed this change as well otherwise it breaks existing numeric test. that means the PR is fully absorbed. let me know if i still miss anything. hope the co-author PR help keep your contribution in record
```
direction_i *= s_i / s_ref
parameter.add_(direction, alpha=-(lr * s_ref))
```

Screenshot 2026-09-21 at 2 49 56 PM
